media/js/firefox/hello/hello-ftu.js

```diff
@@ -0,0 +1,471 @@
+
+// create namespace
+if (typeof Mozilla === 'undefined') {
+    var Mozilla = {};
+}
+
+(function($) {
+    'use strict';
+
+    var HelloFTU = {};
+    var _highlightTimeout;
+    var $defaultCopy = $('.default');
+    var $startCopy = $('.start');
+    var $endCopy = $('.end');
+
+    HelloFTU.highlightSupressed = false;
+
+    // Used for unit testing purposes only as "document.hidden" is read only.
+    HelloFTU.documentHidden = null;
+
+    HelloFTU.tourSource = 'none';
+
+    /**
+     * Initialize Hello FTU.
+     * Only start the tour if the Hello icon is present in the tool pallet.
+     * Given that this tour should only be accessed via the Hello panel itself,
+     * this should always be "true". However, it's always possible someone may
+     * share the URL directly with someone who has disabled or removed Hello.
+     */
+    HelloFTU.init = function() {
+        // Hello is still referred to as 'loop' internally for legacy reasons.
+        Mozilla.UITour.getConfiguration('availableTargets', function (config) {
+            if (config.targets && $.inArray('loop', config.targets) !== -1) {
+                HelloFTU.tourStep = HelloFTU.getInitialTourStep();
+                HelloFTU.setInitialPageState();
+                HelloFTU.bindPageEvents();
+                HelloFTU.bindHelloEvents();
+                HelloFTU.trackTourSource();
+                HelloFTU.showTourStep();
+            }
+        });
+    };
+
+    HelloFTU.getInitialTourStep = function() {
+        // URL query param populated at template level in the view
+        var incomingConversation = HelloFTU.getParameterByName('incomingConversation');
+        var tourStep = 'get-started';
+
+        // set the tour state based on incomingConversation status
+        switch(incomingConversation) {
+        case 'waiting':
+            tourStep = 'conversation-waiting';
+            break;
+        case 'open':
+            tourStep = 'conversation-open';
+            break;
+        }
+
+        return tourStep;
+    };
+
+    HelloFTU.setInitialPageState = function() {
+        if (HelloFTU.tourStep === 'conversation-open') {
+            HelloFTU.showPageState('end');
+            HelloFTU.replaceURLState('open', 'done');
+            HelloFTU.trackGAConversationConnect();
+        } else {
+            HelloFTU.showPageState('start');
+        }
+    };
+
+    /*
+     * Show the Hello panel and trigger callback once opening animation finishes
+     * @param callback (function)
+     */
+    HelloFTU.showHelloPanel = function(callback) {
+        // Only open the info panel if tab is visible
+        if (HelloFTU.documentHidden || document.hidden) {
+            return;
+        }
+        // Make sure loop icon is available before opening Hello panel (bug 1111828)
+        Mozilla.UITour.getConfiguration('availableTargets', function (config) {
+            if (config.targets && $.inArray('loop', config.targets) !== -1) {
```

Comment:
Yeah, this gets called at multiple times in the tour depending on the state, so it's there in the event that the user could remove the Hello button at any time. It's only really costly in that it requires firing a custom event each time, but I think it's better to pay that cost here.